src/crimson/os/seastore/extent_placement_manager.cc

```diff
@@ -0,0 +1,135 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
+// vim: ts=8 sw=2 smarttab expandtab
+
+#include "crimson/os/seastore/extent_placement_manager.h"
+#include "crimson/os/seastore/lba_manager/btree/lba_btree_node_impl.h"
+#include "crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.h"
+#include "crimson/os/seastore/onode_manager/staged-fltree/node_extent_manager/seastore.h"
+#include "crimson/os/seastore/collection_manager/collection_flat_node.h"
+#include "crimson/os/seastore/object_data_handler.h"
+
+namespace crimson::os::seastore {
+
+ExtentAllocWriter::write_ertr::future<>
+ExtentAllocWriter::write(CachedExtent* extent) {
+  bufferlist bl;
+  extent->prepare_write();
+  bl.append(extent->get_bptr());
+  auto segment = extent->get_allocated_segment();
+  assert(extent->extent_writer);
+  assert(segment);
+  auto old_segment_iter = open_segments.begin();
+  if (segment.get() != old_segment_iter->get()) {
+    // segment writes are ordered, so when an opened segment isn't the current
+    // one being written to, it is full and should be closed.
+    (void) open_segments.front()->close();
+    open_segments.erase(old_segment_iter);
+  }
+  return segment->write(extent->get_paddr().offset, bl).safe_then(
+    [this, extent, segment] {
+    extent->extent_writer = nullptr;
+    extent->clear_allocated_segment();
+  });
+}
+
+ExtentAllocWriter::alloc_extent_ertr::future<alloc_t>
+ExtentAllocWriter::alloc(segment_off_t length) {
```

Comment:
Yeah, will correct this part:-)

Reply:
I actually figured that this interface (perhaps not called alloc?) would perform the actual write and return the resulting absolute address for inclusion in the commit transaction within another block (lba tree, etc) as a delta or whole block depending.

Reply:
@athanatos Are you suggesting that the current transaction be splitted into two, one for rewriting extents, the result of which would serve as input to the other transaction in which internal nodes of the lba tree would be modified? :-)

Reply:
@athanatos Do you mean that the actual write be performed in `LBAManager::rewrite_extent()`?:-)

Reply:
@athanatos @cyx1231st I think maybe I should model the current relative paddr mechanism, in which I record the start paddr of the current transaction's rewritten extents in the current transaction, and make internal nodes of the lba tree reference the rewritten extents by a paddr relative to that base paddr. How do you like it?:-)

Reply:
@athanatos @cyx1231st I think I can model the current relative paddr mechinery. To be specific, I think I can:

1. record the id and the base paddrs of extents that are written by the same ExtentAllocWriter in the transaction,
2. make lba tree's internal nodes reference the above extents by the ExtentAllocWriter id and paddrs relative to those base paddrs

How do you like it?:-) Thanks.

Reply:
I think we're missing something here.  We are talking about a new extent located in a segment that *isn't the journal segment*.  Performing a write to that segment has no effect whatsoever on the external state of seastore (no seastore.h level read could detect a difference) *until we commit a journal write referencing it*.  What I'm describing is very similar to bluestore's aio write path -- we perform the write of the new segment to a location, and then commit a journal record referencing it directly.  I see no reason worry about relative addrs here -- the only reason we do that with journal records is to enable us to reference extents within the same atomic write.

Note that this is how it must work -- we cannot atomically write to both the new segment and the journal segment, and we can't write to the journal until the non-journal data is committed.

Reply:
I suppose it needs to be two calls: alloc_ool_extent (alloc out of line extent) returns a CachedExtentRef to be filled in with data.  get_paddr() would need to be a special invalid value (UNKNOWN_ABSOLUTE_EXTENT_MANAGER or something).  A second call (persist_ool_extent perhaps) performs the write to some non-journal location, updates the addr to reflect the actual location, and updates the transaction to add the logical mapping.

Reply:
I think this would be simpler than introducing yet another set of relative addrs.

> A second call (persist_ool_extent perhaps) performs the write to some non-journal location, updates the addr to reflect the actual location, and updates the transaction to add the logical mapping.

I'm wondering about the intense transaction conflict seen by @xxhdx1985126. If this can be implemented after entering `write_pipeline.prepare` with the final confilict checks, and before committing to the journal (maybe add another phase of `ool_submission`?), so that no slow disk write can be wasted by transaction invalidation. What do you think?

Reply:
I think the write itself should occur prior to submit_transaction.  The specific conflict I think you are referring to is simply an artifact of attempting to create all of the pg collections on osd creation in parallel -- they'll conflict on the collection_manager extent.  It's not pertinent to the normal IO pathway.  If conflicts become a problem, we'll look into ways of mitigating that.